pkg/lib/cockpit-upload-helper.ts

```diff
@@ -0,0 +1,133 @@
+    destination: string,
+    stream: ReadableStream,
+    progress?: (bytes_sent: number) => void,
+    signal?: AbortSignal,
+    options?: cockpit.JsonObject
+) {
+    /* While the upload is running, the function is essentially always waiting
+     * inside of a Promise resolved by notify().  Anything that might cause
+     * forward progress (ack messages, close messages, cancellation) will call
+     * notify().  It's up to the main loop body to re-evaluate and respond.
+     */
+    let outstanding = 0; // for flow control
+    let delivered = 0; // for progress reporting
+    let close_message = null as cockpit.JsonObject | null;
+    let notify = () => { };
+
+    if (signal) {
+        signal.throwIfAborted(); // early exit
+        signal.addEventListener('abort', () => { notify() });
+    }
+
+    debug('requesting channel');
+    const channel = cockpit.channel({
+        payload: 'fsreplace1',
+        path: destination,
+        binary: true,
+        'send-acks': 'bytes',
+        ...options,
+    });
+    channel.addEventListener('control', (_ev, message) => {
+        debug('control', message);
+        if (message.command === 'ack') {
+            cockpit.assert(typeof message.bytes === 'number', 'bytes not a number');
+            delivered += message.bytes;
+            if (progress) {
+                debug('progress', delivered);
+                progress(delivered);
+            }
+            outstanding -= message.bytes;
+            debug('outstanding -- to', outstanding);
+            notify();
+        }
+    });
+    channel.addEventListener('close', (_ev, message) => {
+        debug('close', message);
+        close_message = message;
+        notify();
+    });
+
+    try {
+        debug('starting file send');
+        const reader = stream.getReader({ mode: 'byob' }); // byob to choose the block size
+        let eof = false;
+
+        // eslint-disable-next-line no-unmodified-loop-condition
+        while (!close_message) {
+            /* We do the following steps for as long as the channel is open:
+             *  - while we have data to write, and room for it, write
+             *  - then wait for an incoming message and start over
+             */
+            while (!eof && outstanding < FLOW_WINDOW) {
+                const { done, value } = await reader.read(new Uint8Array(BLOCK_SIZE));
+                if (done) {
+                    debug('sending done');
+                    channel.control({ command: 'done' });
+                    eof = true;
+                } else {
+                    debug('sending', value.length, 'bytes');
+                    channel.send(value);
+                    outstanding += value.length;
+                    debug('outstanding ++ to', outstanding);
+                }
+            }
+
+            debug('...sleeping', outstanding, 'of', FLOW_WINDOW, 'eof', eof);
+            await new Promise<void>(resolve => { notify = resolve });
+            if (signal) {
```

Comment:
Now we would only abort if all messages are in flight and the promise above is resolved